### DISCUSSION TITLE
Tracing RavenDB requests

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#13334] [Q&A] [answered]
Hi everybody,

I am trying to add tracing to all requests that are going to RavenDB.

I found the following idea
https://github.com/ravendb/ravendb/discussions/12432
where it is recommended to use the `OnBeforeRequest` and `OnSucceedRequest` / `OnFailedRequest` to get timing metrics.

However, for tracing the tracing library APIs don't really go well along with this concept, as generally a disposable object is used in a block scope fashion for instrumentation.

see e.g. https://docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#optional-add-additional-activities

```csharp
        static async Task StepOne()
        {
            using (Activity activity = source.StartActivity("StepOne"))
            {
                await Task.Delay(500);
            }
        }
```

same applies to GCP tracing lib
https://googleapis.dev/dotnet/Google.Cloud.Diagnostics.AspNetCore/4.3.0/index.html#configuration-2

I think a possible way around it would be using the IHttpClientFactory, but it seems RavenDB is creating the client directly using `new HttpClient(...)`
https://github.com/ravendb/ravendb/blob/f773b7da6e8afccad88549ee528fd1165878baa1/src/Raven.Client/Http/RequestExecutor.cs#L1903

If the factory would be used, like recommended here
https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Diagnostics.Common/latest#trace-outgoing-http-requests
https://github.com/googleapis/google-cloud-dotnet/blob/main/apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/HttpClientBuilderExtensions.cs
we could register a trace handler like that:

```csharp
public static IHostBuilder CreateHostBuilder() =>
    new HostBuilder()
        .ConfigureServices(services =>
        {
            // Replace ProjectId with your Google Cloud Project ID.
            services.AddGoogleTrace(new TraceServiceOptions { ProjectId = ProjectId });

            // Register an HttpClient for outgoing requests.
            services.AddHttpClient("tracesOutgoing")
                // The next call guarantees that trace information is propagated for outgoing
                // requests that are already being traced.
                .AddOutgoingGoogleTraceHandler();

            // Register other services here if you need them.
        });

----
    IHttpClientFactory clientFactory = host.Services.GetRequiredService<IHttpClientFactory>();
    var httpClient = clientFactory.CreateClient("tracesOutgoing");
    // Any code that makes outgoing requests.
    var response = await httpClient.GetAsync("http://weather.com/");
```

Would be great if anybody has some recommendations on how to hook up those two concepts!

**Comment by vlko** [comment]
Here is our integration of RavenDB request tracing with HttpDiagnosticSourceSubscriber to MiniProfiler https://gist.github.com/vlko/708c046c547c6684c547154ebaa74f1f

**Reply by danielwinkler** [reply]
Thank you very much for sharing your code!
I will go with Ayende's proposal first, but it's great to see how to use reflection in case where we don't have the necessary hooks at all.

**Comment by ayende** [comment, marked as answer]
You can handle this like so, no?


```
store.OnBeforeRequest += (sender, e) =>
{
   e.Request.Properties["MyTracking"] = source.StartActivity("StepOne");
};

store.OnSucceedRequest += (sender, e) =>
{
    if(e.Request.Properties["MyTracking"]  is IDisposable d)
          d.Dispose();
};
```

**Reply by danielwinkler** [reply]
awesome! thank you very much!
